src/include/spinlock.h

```diff
@@ -18,75 +18,141 @@
+  // For details please refer to:
+  //  * "Long Duration Spin-wait Loops on Hyper-Threading
+  //     Technology Enabled Intel Processors",
+  //  * "Benefitting Power and Performance Sleep Loops".
+  _mm_pause();
+#endif
+}
 
-inline void spin_lock(std::atomic_flag& lock);
-inline void spin_unlock(std::atomic_flag& lock);
-inline void spin_lock(ceph::spinlock& lock);
-inline void spin_unlock(ceph::spinlock& lock);
 
-/* A pre-packaged spinlock type modelling BasicLockable: */
-class spinlock final
+template <class MutexT, std::size_t MaxTriesV = 64>
+struct trying_guard final
 {
-  std::atomic_flag af = ATOMIC_FLAG_INIT;
-
-  public:
-  void lock() {
-    ceph::spin_lock(af);
-  }
- 
-  void unlock() noexcept {
-    ceph::spin_unlock(af);
+  typedef MutexT mutex_type;
+
+  trying_guard(mutex_type& m)
+    : m(m)
+  {
+    if (likely(m.try_lock())) {
+      return;
+    }
+
+    std::size_t tries = MaxTriesV;
+    do {
+      emit_pause();
+    } while (!m.try_lock() && --tries > 0);
+
+    if (!tries) {
+      m.lock();
+    }
   }
-};
 
-// Free functions:
-inline void spin_lock(std::atomic_flag& lock)
-{
- while(lock.test_and_set(std::memory_order_acquire))
-  ;
-}
+  trying_guard(const trying_guard&) = delete;
 
-inline void spin_unlock(std::atomic_flag& lock)
-{
- lock.clear(std::memory_order_release);
-}
+  ~trying_guard()
+  {
+    m.unlock();
+  }
 
-inline void spin_lock(std::atomic_flag *lock)
-{
- spin_lock(*lock);
-}
+private:
+  mutex_type& m;
+};
 
-inline void spin_unlock(std::atomic_flag *lock)
-{
- spin_unlock(*lock);
-}
 
-inline void spin_lock(ceph::spinlock& lock)
+/* A pre-packaged spinlock type modelling BasicLockable: */
+template <std::size_t MaxTriesV = 32>
+class spinlock final
 {
- lock.lock();
-}
+  // Not using atomic_flag anymore because it doesn't
+  // provide the load nor store operation.
+  std::atomic_bool locked = false;
+
+  // In contrast to atomic_flag, atomic_bool might be
+  // implemented on top of e.g. mutex. However, it is
+  // very unlikely we'll face such situation in Ceph.
+  static_assert(std::atomic_bool::is_always_lock_free);
+
+public:
+  void lock();
+  void unlock() noexcept;
+};
 
-inline void spin_unlock(ceph::spinlock& lock)
+template <std::size_t MaxTriesV>
+inline void spinlock<MaxTriesV>::lock()
 {
- lock.unlock();
-}
+  bool expected = false;
+  if (likely(locked.compare_exchange_weak(expected, true,
+                                          std::memory_order_acquire,
+                                          std::memory_order_relaxed))) {
+    return;
+  }
 
-inline void spin_lock(ceph::spinlock *lock)
-{
- spin_lock(*lock);
+  // The contention part.
+  do {
+    std::size_t tries = 0;
+
+    // There is no need to constantly try LOCK CMPXCHG and enforce
+    // x86 CPUs to do costly mem fencing. The algorithm comes from
+    // NPTL's pthread_spin_lock() implementation of glibc.
+    do {
+      emit_pause();
+
+      if constexpr (MaxTriesV) if (++tries == MaxTriesV) {
```

Comment:
How long is a pause valid for? We try not to hold spin locks for a very long time but if we're sleeping after 32 cpu instructions, that doesn't seem long enough :o

Reply:
As usually: it depends. ;-) On Skylake and newer microarchitectures `PAUSE` takes around 140 cycles in comparison to ~10 for previous ones. Those numbers came from page 326 of the [Intel® 64 and IA-32 Architectures Optimization Reference Manual](https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-optimization-manual.pdf).

Reply:
Hmm, okay, 320-4480 cycles is definitely *more* plausible. I'm just concerned that while most of our spinlocks are incredibly short, some of them do involve traversing std::maps or whatever that I'm sure are reasonably efficient but might be large enough to take more than 32 instructions. ;)